File: webmcp/src/query.js
// Component-query parser — a compatible subset of the sightmap CLI's
// component-query DSL (see the sightmap-browser skill):
//
//   ProductCard                          bare component name
//   ProductCard[name^="Weber"]           predicate on an extracted property
//   ProductCard[name*="weber" i]         substring, case-insensitive
//   LoginForm UserNameInput              descendant chain (target = last)
//   FulfillmentTileButton#1              occurrence index (0-based)
//   css:.plp-price-module a              raw-CSS escape hatch (not in the CLI)
//
// Operators: `=` exact, `^=` prefix, `*=` substring; trailing ` i` flag for
// case-insensitive. Whitespace is a descendant combinator; there is no `>`.
// Predicate values may carry `{param}` templates, interpolated at runtime.

const NAME_RE = /^[A-Za-z][A-Za-z0-9_-]*/;

function parseLink(text) {
  const nameMatch = text.match(NAME_RE);
  if (!nameMatch) throw new Error(`expected a component name at "${text}"`);
  const link = { name: nameMatch[0], preds: [], index: null };
  let rest = text.slice(nameMatch[0].length);
  while (rest.startsWith("[")) {
    const end = findPredEnd(rest);
    link.preds.push(parsePred(rest.slice(1, end)));
    rest = rest.slice(end + 1);
  }
  if (rest.startsWith("#")) {
    const m = rest.match(/^#(\d+)/);
    if (!m) throw new Error(`expected an index after "#" in "${text}"`);
    link.index = parseInt(m[1], 10);
    rest = rest.slice(m[0].length);
  }
  if (rest.length > 0) {
    throw new Error(`unexpected trailing "${rest}" in query link "${text}"`);
  }
  return link;
}

function findPredEnd(s) {
  // s starts with '['; find the matching ']' respecting quotes.
  let quote = null;
  for (let i = 1; i < s.length; i++) {
    const c = s[i];
    if (quote) {
      if (c === quote) quote = null;
    } else if (c === '"' || c === "'") {
      quote = c;
    } else if (c === "]") {
      return i;
    }
  }
  throw new Error(`unterminated predicate in "${s}"`);
}

function parsePred(body) {
  const m = body.match(/^([a-z][a-z0-9_]*)\s*(\^=|\*=|=)\s*([\s\S]*)$/);
  if (!m) throw new Error(`cannot parse predicate "[${body}]"`);
  let value = m[3].trim();
  let ci = false;
  // A trailing ` i` flag counts when what precedes it is a complete quoted
  // string, or when the whole value is bare (an unquoted value can't itself
  // end in ` i` and mean the letter).
  const noFlag = value.replace(/\s+i$/, "");
  if (noFlag !== value && (isQuoted(noFlag) || !isQuoted(value))) {
    ci = true;
    value = noFlag;
  }
  if (isQuoted(value)) value = value.slice(1, -1);
  return { prop: m[1], op: m[2], value, ci };
}

function isQuoted(v) {
  return (
    (v.startsWith('"') && v.endsWith('"') && v.length >= 2) ||
    (v.startsWith("'") && v.endsWith("'") && v.length >= 2)
  );
}

// splitLinks splits on whitespace that is outside [...] predicates.
function splitLinks(q) {
  const parts = [];
  let depth = 0;
  let quote = null;
  let cur = "";
  for (const c of q) {
    if (quote) {
      cur += c;
      if (c === quote) quote = null;
    } else if (c === '"' || c === "'") {
      cur += c;
      quote = c;
    } else if (c === "[") {
      depth++;
      cur += c;
    } else if (c === "]") {
      depth--;
      cur += c;
    } else if (/\s/.test(c) && depth === 0) {
      if (cur) parts.push(cur);
      cur = "";
    } else {
      cur += c;
    }
  }
  if (cur) parts.push(cur);
  return parts;
}

// parseQuery parses a component query (or css: escape) into an IR the
// compiler resolves against the corpus.
function parseQuery(q) {
  const text = String(q).trim();
  if (text.startsWith("css:")) {
    const selector = text.slice(4).trim();
    if (!selector) throw new Error(`empty selector in "${q}"`);
    return { kind: "css", selector };
  }
  if (!text) throw new Error("empty component query");
  if (/\s>\s|>/.test(text.replace(/\[[^\]]*\]/g, ""))) {
    throw new Error(
      `"${q}": the child combinator ">" is not supported — whitespace is a descendant combinator`,
    );
  }
  const links = splitLinks(text).map(parseLink);
  if (links.length === 0) throw new Error("empty component query");
  return { kind: "query", links };
}

module.exports = { parseQuery };

File: go/webmcp/query.go
package webmcp

// Component-query parser — a direct port of webmcp/src/query.js. See that
// file (and the sightmap-browser skill) for the DSL this accepts.

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type queryPred struct {
	Prop  string
	Op    string
	Value string
	CI    bool
}

type queryLink struct {
	Name  string
	Preds []queryPred
	Index *int // nil when absent
}

type parsedQuery struct {
	Kind     string // "query" | "css"
	Selector string // css only
	Links    []queryLink
}

var queryNameRe = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9_-]*`)
var predBodyRe = regexp.MustCompile(`^([a-z][a-z0-9_]*)\s*(\^=|\*=|=)\s*([\s\S]*)$`)
var indexRe = regexp.MustCompile(`^#(\d+)`)
var trailingFlagRe = regexp.MustCompile(`\s+i$`)

func isQuoted(v string) bool {
	return len(v) >= 2 &&
		((strings.HasPrefix(v, `"`) && strings.HasSuffix(v, `"`)) ||
			(strings.HasPrefix(v, `'`) && strings.HasSuffix(v, `'`)))
}

func parsePred(body string) (queryPred, error) {
	m := predBodyRe.FindStringSubmatch(body)
	if m == nil {
		return queryPred{}, fmt.Errorf("cannot parse predicate \"[%s]\"", body)
	}
	value := strings.TrimSpace(m[3])
	ci := false
	noFlag := trailingFlagRe.ReplaceAllString(value, "")
	if noFlag != value && (isQuoted(noFlag) || !isQuoted(value)) {
		ci = true
		value = noFlag
	}
	if isQuoted(value) {
		value = value[1 : len(value)-1]
	}
	return queryPred{Prop: m[1], Op: m[2], Value: value, CI: ci}, nil
}

func findPredEnd(s string) (int, error) {
	var quote byte
	for i := 1; i < len(s); i++ {
		c := s[i]
		if quote != 0 {
			if c == quote {
				quote = 0
			}
		} else if c == '"' || c == '\'' {
			quote = c
		} else if c == ']' {
			return i, nil
		}
	}
	return 0, fmt.Errorf("unterminated predicate in %q", s)
}

func parseLink(text string) (queryLink, error) {
	name := queryNameRe.FindString(text)
	if name == "" {
		return queryLink{}, fmt.Errorf("expected a component name at %q", text)
	}
	link := queryLink{Name: name}
	rest := text[len(name):]
	for strings.HasPrefix(rest, "[") {
		end, err := findPredEnd(rest)
		if err != nil {
			return queryLink{}, err
		}
		pred, err := parsePred(rest[1:end])
		if err != nil {
			return queryLink{}, err
		}
		link.Preds = append(link.Preds, pred)
		rest = rest[end+1:]
	}
	if strings.HasPrefix(rest, "#") {
		m := indexRe.FindStringSubmatch(rest)
		if m == nil {
			return queryLink{}, fmt.Errorf("expected an index after \"#\" in %q", text)
		}
		n, _ := strconv.Atoi(m[1])
		link.Index = &n
		rest = rest[len(m[0]):]
	}
	if rest != "" {
		return queryLink{}, fmt.Errorf("unexpected trailing %q in query link %q", rest, text)
	}
	return link, nil
}

// splitLinks splits on whitespace outside [...] predicates.
func splitLinks(q string) []string {
	var parts []string
	depth := 0
	var quote byte
	var cur strings.Builder
	for i := 0; i < len(q); i++ {
		c := q[i]
		switch {
		case quote != 0:
			cur.WriteByte(c)
			if c == quote {
				quote = 0
			}
		case c == '"' || c == '\'':
			cur.WriteByte(c)
			quote = c
		case c == '[':
			depth++
			cur.WriteByte(c)
		case c == ']':
			depth--
			cur.WriteByte(c)
		case (c == ' ' || c == '\t' || c == '\n' || c == '\r') && depth == 0:
			if cur.Len() > 0 {
				parts = append(parts, cur.String())
				cur.Reset()
			}
		default:
			cur.WriteByte(c)
		}
	}
	if cur.Len() > 0 {
		parts = append(parts, cur.String())
	}
	return parts
}

var predStripRe = regexp.MustCompile(`\[[^\]]*\]`)

func parseQuery(q string) (parsedQuery, error) {
	text := strings.TrimSpace(q)
	if strings.HasPrefix(text, "css:") {
		selector := strings.TrimSpace(text[4:])
		if selector == "" {
			return parsedQuery{}, fmt.Errorf("empty selector in %q", q)
		}
		return parsedQuery{Kind: "css", Selector: selector}, nil
	}
	if text == "" {
		return parsedQuery{}, fmt.Errorf("empty component query")
	}
	if strings.Contains(predStripRe.ReplaceAllString(text, ""), ">") {
		return parsedQuery{}, fmt.Errorf(
			"%q: the child combinator \">\" is not supported — whitespace is a descendant combinator", q)
	}
	var links []queryLink
	for _, part := range splitLinks(text) {
		link, err := parseLink(part)
		if err != nil {
			return parsedQuery{}, err
		}
		links = append(links, link)
	}
	if len(links) == 0 {
		return parsedQuery{}, fmt.Errorf("empty component query")
	}
	return parsedQuery{Kind: "query", Links: links}, nil
}
